x/superfluid/keeper/stake.go

```diff
@@ -656,3 +659,198 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+// - gamm share denom is in correct format
+func (k Keeper) validateUnbondConvertAndStake(ctx sdk.Context, sharesToStake sdk.Coin) (poolIdLeaving uint64, err error) {
+	// Defense in depth, ensuring the sharesToMigrate contains gamm pool share prefix.
+	if !strings.HasPrefix(sharesToStake.Denom, gammtypes.GAMMTokenPrefix) {
+		return 0, types.SharesToMigrateDenomPrefixError{Denom: sharesToStake.Denom, ExpectedDenomPrefix: gammtypes.GAMMTokenPrefix}
+	}
+
+	// Get the balancer poolId by parsing the gamm share denom.
+	poolIdLeaving, err = gammtypes.GetPoolIdFromShareDenom(sharesToStake.Denom)
+	if err != nil {
+		return 0, err
+	}
+
+	return poolIdLeaving, nil
+}
+
+// convertGammSharesToOsmoAndStake converts given gamm shares to osmo by swapping in the given pool
+// then stakes it to the designated validator.
+// minAmtToStake works as slippage bound, and would error if total amount being staked is less than min amount to stake.
+// If valAddr is empty, we attempt to get staking preference from valset-pref module and stake to the given validator.
+func (k Keeper) convertGammSharesToOsmoAndStake(
+	ctx sdk.Context,
+	sender sdk.AccAddress, valAddr string,
+	poolIdLeaving uint64, exitCoins sdk.Coins, minAmtToStake sdk.Int,
+) (totalAmtCoverted sdk.Int, shares sdk.Dec, err error) {
+	var nonOsmoCoins sdk.Coins
+	bondDenom := k.sk.BondDenom(ctx)
+
+	// from the exit coins, separate non-bond denom and bond denom.
+	for _, exitCoin := range exitCoins {
+		// if coin is not uosmo, add it to non-osmo Coins
+		if exitCoin.Denom != bondDenom {
+			nonOsmoCoins = append(nonOsmoCoins, exitCoin)
+		}
+	}
+	originalBondDenomAmt := exitCoins.AmountOf(bondDenom)
+
+	// track how much non-uosmo tokens we have converted to uosmo
+	totalAmtCoverted = sdk.ZeroInt()
+
+	// iterate over non-bond denom coins and swap them into bond denom
+	for _, coinToConvert := range nonOsmoCoins {
+		tokenOutAmt, err := k.pmk.SwapExactAmountIn(ctx, sender, poolIdLeaving, coinToConvert, bondDenom, sdk.ZeroInt())
+		if err != nil {
+			return sdk.ZeroInt(), sdk.ZeroDec(), err
+		}
+
+		totalAmtCoverted = totalAmtCoverted.Add(tokenOutAmt)
+	}
+
+	// add the converted amount with the amount of osmo from exit coin to get total amount we would be staking
+	totalAmtToStake := originalBondDenomAmt.Add(totalAmtCoverted)
+
+	// check if the total amount to stake after all conversion is greater than provided min amount to stake
+	if totalAmtToStake.LT(minAmtToStake) {
+		return sdk.ZeroInt(), sdk.ZeroDec(), types.TokenConvertedLessThenDesiredStakeError{
+			ActualTotalAmtToStake:   totalAmtToStake,
+			ExpectedTotalAmtToStake: minAmtToStake,
+		}
+	}
+
+	var val stakingtypes.Validator
+	// if given valAddr is empty, we use delegation preference given from valset-pref module.
+	if valAddr == "" {
+		delegationPref, err := k.vspk.GetDelegationPreferences(ctx, sender.String())
+		if err != nil {
+			return sdk.ZeroInt(), sdk.ZeroDec(), err
+		}
+		if len(delegationPref.Preferences) != 1 {
```

Comment:
What do folks think about this? Would this be ok? If val set pref returns multiple validators, made it so that it just errored here cc @sunnya97

Reply:
The flow imo should be:
- If valAddr provided, single delegate
- If valAddr not provided, and valset exists, valsetpref.Delegate
- If valAddr not provided, valset does not exist, and superfluid'd, use superfluid validator
- Else: Error